### PR TITLE
CI (Buildkite, GHA): Allow any user with triage or commit permissions to retry all failed Buildkite jobs

### PR DESCRIPTION
**Pull request by DilumAluthge**
This PR adds a GitHub Actions workflow that allows any user with triage or commit permissions to retry all failed Buildkite jobs by adding the `Buildkite - retry failed jobs` label.

The key points are:
1. Any user with triage or commit permissions can do this.
2. Users do not need to create a Buildkite account. They just add the `Buildkite - retry failed jobs` label to the PR.
2. It only reruns failed Buildkite jobs. It doesn't rerun any passed Buildkite jobs, which helps conserve CI resources.

TODO before merging:
- [x] Remove the trailing whitespace
- [x] Change `pull_request` to `pull_request_target`

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Alright, this is working correctly, and it is now ready for review. @staticfloat 

Note that there are two TODOs that should be done right before merging.

File: .github/workflows/retry.yml
Comment (staticfloat) [inline .github/workflows/retry.yml:29]: Can't someone just edit this file and open a PR with a modified `.yml` file?

File: .github/workflows/retry.yml
Comment (staticfloat) [inline .github/workflows/retry.yml:36]: Are these TODOs still relevant?

File: .github/workflows/retry.yml
Comment (staticfloat) [inline .github/workflows/retry.yml:47]: Is this still relevant?  I thought this PR was buildkite-specific and has nothing to do with buildbot.

File: .github/workflows/retry.yml
Comment (DilumAluthge) [reply .github/workflows/retry.yml:29]: No. If someone (who does not have commit access) makes a malicious PR from a fork, then:
- The `GITHUB_TOKEN` will only have read-only access.
- No other secrets will be available. So, for example, the `BUILDKITE_API_TOKEN_RETRY` secret will not be available.

There are two types of events to consider here:
1. The `pull_request` event
2. The `pull_request_target` event

When someone makes a PR from a fork, the `pull_request` event will:
- Use the copy of the workflow file in their PR
- Only have a read-only `GITHUB_TOKEN`
- Not have access to any other secrets

So for the `pull_request` event, yes the workflow file has malicious content, but the `GITHUB_TOKEN` is read-only, and there is no access to any other secrets, so the only thing that the attacker can do is make read calls to the GitHub API.

When someone makes a PR from a fork, the `pull_request_target` event will:
- Use the copy of the workflow file in the `master` branch
- Have a `GITHUB_TOKEN` with write access
- Have access to secrets (including but not limited to the `BUILDKITE_API_TOKEN_RETRY` secret)

So, in the `pull_request_target` event, the job that runs has write access and access to secrets. However, the workflow file that is being used is the one from the `master` branch. So as long as we don't merge the attacker's PR, their malicious code will not be run.

Does that make sense?

Comment (DilumAluthge) [reply .github/workflows/retry.yml:29]: The main purpose of the `Therefore, for security reasons, we do not checkout any code in this workflow` comment is to remind me (and other committers) that we should not merge any PR that checks out code as part of the `pull_request_target` event.

Basically, it's a reminder to follow the security guidelines for the `pull_request_target` event discussed here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

Comment (staticfloat) [reply .github/workflows/retry.yml:29]: Perfect!  I love the GH integration here.

File: .github/workflows/retry.yml
Comment (staticfloat) [inline .github/workflows/retry.yml:77]: Whoever said Julia didn't make a good shell scripting language? :P

File: .github/workflows/retry.yml
Comment (DilumAluthge) [reply .github/workflows/retry.yml:36]: This deserves some explanation.

Okay, so if you look at the workflow file currently, you'll see it runs on two types of event:
1. When the `Buildkite - retry failed jobs` label is added to a PR.
2. When a PR is reopened after being closed.

Once we have fully migrated to Buildkite, I will remove option 1 and only have option 2.

In the current situation, in which we have both Buildbot and Buildkite, here is what happens in each event:
1. Label is added: In this event, all of the failed Buildkite jobs will be rerun. None of the passed Buildkite jobs will be rerun. **Nothing happens to the Buildbot jobs.**
2. PR is reopened after being closed: In this event, all of the failed Buildkite jobs will be rerun. None of the passed Buildkite jobs will be rerun. **All of the Buildbot jobs (both passed and failed) will be rerun.**

Once we have migrated everything to Buildkite, and we no longer have anything on Buildbot, here is what happens in each event:
1. Label is added: In this event, all of the failed Buildkite jobs will be rerun.
2. PR is reopened after being closed: In this event, all of the failed Buildkite jobs will be rerun.

So, currently (in the Buildbot + Buildkite world), the "Label is added" event is preferable, because it doesn't rerun the passed Buildbot jobs, and thus conserves CI resources.

But in the future Buildkite-only world, the "Label is added" event is exactly identical in effect to the "PR is reopened after being closed" event. And I think, personally, that telling someone to "close and reopen a PR, and it will only rerun failed CI" is a much more user-friendly interface than this label business.

So, once we are in the Buildkite-only world, I want to completely remove the label trigger, and only have the "close and reopen the PR" trigger, because I think it's easier for people to use.

So these TODOs are here to remind me to get rid of the label stuff once we are Buildkite-only.

What do you think?

**Review by staticfloat — APPROVED** [review]
With your explanations, this now makes sense, but I really dislike the feeling of embedding a huge amount of code within the `.yml` files.  I understand you want to avoid using any checked-out code so that it's harder for people to jam malicious stuff into these PRs/the main repository.

I feel like the `cryptic-buildkite-plugin` approach solves this but on a different point in the solution space that is slightly more work for maintainers in updating, but provides a more "natural" experience (e.g. you can have this chunk of Julia code living as actual `.jl` files without fear).  I don't have strong enough feelings to say "no" to this pull request, but I do think we should try and make something like `cryptic` convenient enough for people like you that you can use that without fear, and just use it instead of jamming more and more code into `.yml` files in the future.

**Comment by DilumAluthge** [comment]
Is the only issue the fact that we have a lot of Julia code in the YAML file? If so, that's actually not a hard thing to fix. I can make a composite action in the JuliaLang organization, and then we can use that composite action (by full commit hash, for security) in the YAML file here.

**Comment by DilumAluthge** [comment]
>  I do think we should try and make something like `cryptic` convenient enough for people like you that you can use that without fear

Yeah, completely independent of this PR, it would be nice for us to make `cryptic` a little smoother to use.

Also, right now, `cryptic` is only for Buildkite, but that doesn't really work here, because we want to run this workflow on very specific triggers (e.g. label is added, but only if the label is this specific label), and Buildkite webhooks don't have that level of granularity.

So it could be cool if we had a `cryptic`-type plugin for GitHub Actions.

**Comment by DilumAluthge** [comment]
> Is the only issue the fact that we have a lot of Julia code in the YAML file? If so, that's actually not a hard thing to fix. I can make a composite action in the JuliaLang organization, and then we can use that composite action (by full commit hash, for security) in the YAML file here.

If we go with that approach, the YAML file in this repo simplifies down to something like this:
```yaml
# Please ping @DilumAluthge when making any changes to this file.

# Here are some steps that we take in this workflow file for security reasons:
# 1. We do not checkout any code.
# 2. We do not run any external actions.
# 3. We do not give the `GITHUB_TOKEN` any permissions.
# 4. We only give the Buildkite API token (`BUILDKITE_API_TOKEN_RETRY`) the minimum necessary
#    set of permissions.

# Important note to Buildkite maintainers:
# In order to make this work, you need to tell Buildkite that it should NOT create a brand-new
# build when someone closes and reopens a pull request. To do so:
# 1. Go to the relevant pipeline (e.g. https://buildkite.com/julialang/julia-master).
# 2. Click on the "Pipeline Settings" button.
# 3. In the left sidebar, under "Pipeline Settings", click on "GitHub".
# 4. In the "GitHub Settings", under "Build Pull Requests", make sure that the "Skip pull
#    request builds for existing commits" checkbox is checked. This is the setting that tells
#    Buildkite that it should NOT create a brand-new build when someone closes and reopens a
#    pull request.
# 5. At the bottom of the page, click the "Save GitHub Settings" button.

name: Retry Failed Buildkite Jobs
on:
  pull_request:
    types: [ reopened, labeled ]
permissions:
  statuses: none
jobs:
  retry:
    name: retry
    runs-on: ubuntu-latest
    if: github.repository == 'JuliaLang/julia' && (github.event.label.name == 'Buildkite - retry failed jobs' || github.event.action == 'reopened')
    steps:
      - uses: JuliaLang/retry-buildkite@my_full_commit_hash
        with:
          buildkite_token: ${{ secrets.BUILDKITE_API_TOKEN_RETRY }}
          pr_number: ${{github.event.number}}
```

**Comment by DilumAluthge** [comment]
@staticfloat I've moved all of the Julia code to a composite action (https://github.com/JuliaLang/retry-buildkite), that we then `use` in the YAML workflow file in this repository.

**Comment by DilumAluthge** [comment]
Also, by putting this functionality into a composite action, we can use it on other repos.

It will work on any repo that uses Buildkite. Just add a Buildkite API token as a GitHub Secret on the repo.

**Comment by DilumAluthge** [comment]
The two TODOs are done.

**Pull request merged**
